### PR TITLE
Officer permissions feature

### PR DESCRIPTION
**Pull request by dmartin-gh**
These changes add a "Guild Permissions" section to the guild `/update` page. Only the guild leader can modify these settings, which default to edit/delete privileges enabled for ranks 0 and 1. Removing the permissions for rank 0 is not allowed, of course. Hovering over the rank member count will give you a sample of up to 5 characters in that rank, sorted by level then name.

screen shot 2015-12-20 at 6 24 01 pm

All edit and delete permissions are checked in the client and server. If you don't have a particular permission, that icon (the pencil or the trash can) will not render for you. In addition, 8d918b5 resolves a bug where the `delete:characterAd` handler wasn't checking if you own the character.

The behavior of the accounts page is less than perfect with these changes. Currently anyone can add a guild they are a member of. This always calls `put:guildAd`, which only officers can now do. If I change this to something like `put:guildId` to simply register that you are part of that guild, the ad is not guaranteed to exist, so the guild doesn't appear in your list afterwards. Since the guild list in the accounts page currently represents guilds with _active_ ads, I left the call as `put:guildAd` and you now get a `GUILD_NOT_OFFICER_ERROR`. Once the guild is in the database and you log in again, the `userService.updateGuildsId()` function is called and you'll be added to the guild record, causing it to appear in your list _if_ they have an active ad.

Long story short, if you are not an officer of the guild it may come and go from your list based on the status of the ad and you have no control over it. It would be nice if the list represented the guilds you are in and not which ones currently have active advertisements.

Also, I'm new to a lot of the packages used in WarcraftLFG (like Angular.js) so if there are better ways to do things please let me know!

### CODE REVIEW

**Comment by dmartin-gh** [comment]
GitHub trick: if you add `?w=1` to a changeset page, it will ignore whitespace changes. This is particularly useful when trying to figure out what I actually changed in `client/app/guild/guild-update.html`

https://github.com/warcraftlfg/warcraft-lfg/pull/222/files?w=1

Unfortunately it looks like they disable line-by-line commenting in this mode for pull requests.

File: client/app/guild/guild.js [hunk 1]
```diff
@@ -76,14 +76,51 @@
...
+                $scope.$parent.loading = false;
+                $scope.guildRank = rank;
+
+                if (rank === 0) {
+                    // This is the guild leader, put the rank permissions in an easier form for table rendering
+                    var perms = $scope.guildRankPerms = [];
+                    var isOfRank = function(i) {
+                        return function (member) { return member.rank === i; };
+                    };
+                    for (var i = 0; i < 10; i++) {
```
Comment (dmartin-gh) [inline client/app/guild/guild.js:97]: Hard-coded `10` here for number of ranks including the GM. Not sure if we want to parameterize this or where it would go, it's unlikely they'll change the number of guild ranks.

File: server/app/services/userService.js [hunk 2]
```diff
@@ -61,6 +61,85 @@ module.exports.getCharacters = function(region,id,callback){
     });
 };
 
+module.exports.getGuildRank = function (id,region,realm,name,callback){
```
Comment (dmartin-gh) [inline server/app/services/userService.js:64]: This function could have also hit bnet directly, but it will be called more often than `isMember` and `isOwner` are currently called and I didn't want anything tied directly to a page refresh making spurious requests to bnet. As such, rank permissions are evaluated based on WarcraftLFG's view of a guild in MongoDB.

**Comment by bmagic** [comment]
For me it's ok. I talk with the team and I merge.
We are using slack, I organize and clean chans & I will send you a request if you want discuss with us.

**Comment by dmartin-gh** [comment]
Sounds good. My email is frostedcheerios@gmail.com.

**Comment by bmagic** [comment]
It's online.

I'have juste change the defaut permission to everyone can edit / create / delete until an officier change this.
